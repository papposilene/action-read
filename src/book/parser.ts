import {exportVariable, setFailed} from "@actions/core";
import {isIsbn, isDate} from "../utils/globals";

export default function parseTitle(title: string): {
    bookIsbn?: string;
    date: string;
} {
    const split = title.split(" ");
    const bookIsbn = isIsbn(split[0]) ? split[0] : undefined;
    if (!bookIsbn) setFailed(`ISBN is not valid: ${title}`);
    const date = isDate(split[1])
        ? split[1]
        : new Date().toISOString().slice(0, 10);
    exportVariable("DateRead", date);
    return {
        bookIsbn,
        date
    };
}
