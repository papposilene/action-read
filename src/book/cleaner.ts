import {removeWrappedQuotes} from "../utils/globals";
import {BookOptions, Book} from "./getter";

export type CleanBook = {
    dateFinished: string;
    title?: string;
    authors?: string[];
    publishedDate?: string;
    description?: string;
    industryIdentifiers?: { type?: string; indentifier?: string }[];
    categories?: string[];
    pageCount?: number;
    printType?: string;
    imageLinks?: {
        smallThumbnail?: string;
        thumbnail?: string;
    };
    language?: string;
    canonicalVolumeLink?: string;
    isbn: string;
    notes?: string;
};

export default function cleanBook(options: BookOptions, book: Book): CleanBook {
    const {date, body, bookIsbn} = options;
    return {
        isbn: bookIsbn,
        dateFinished: date,
        ...(body && {notes: body}),
        ...("title" in book && {title: book.title}),
        ...("authors" in book && {
            authors: book.authors
        }),
        ...("publishedDate" in book && {publishedDate: book.publishedDate}),
        ...("description" in book && {
            description: removeWrappedQuotes(book.description)
        }),
        ...("industryIdentifiers" in book && {
            industryIdentifiers: book.industryIdentifiers
        }),
        ...("pageCount" in book && {pageCount: book.pageCount}),
        ...("printType" in book && {printType: book.printType}),
        ...("categories" in book && {categories: book.categories}),
        ...("imageLinks" in book && {
            imageLinks: {
                ...("smallThumbnail" in book.imageLinks && {
                    smallThumbnail: book.imageLinks.smallThumbnail.replace(
                        "http:",
                        "https:"
                    )
                }),
                ...("thumbnail" in book.imageLinks && {
                    thumbnail: book.imageLinks.thumbnail.replace("http:", "https:")
                })
            }
        }),
        ...("language" in book && {language: book.language}),
        ...("canonicalVolumeLink" in book && {
            canonicalVolumeLink: book.canonicalVolumeLink
        })
    };
}
