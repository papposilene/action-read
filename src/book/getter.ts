import {exportVariable} from "@actions/core";
import isbn from "node-isbn";
import addBook from "./setter";
import {CleanBook} from "./cleaner";

export type Book = {
    title: string;
    authors: string[];
    publisher: string;
    publishedDate: string;
    description: string;
    industryIdentifiers: {
        type: string;
        identifier: string;
    }[];
    readingModes: {
        text: boolean;
        image: boolean;
    };
    pageCount: number;
    printType: string;
    categories: string[];
    averageRating: number;
    ratingsCount: number;
    maturityRating: string;
    allowAnonLogging: boolean;
    contentVersion: string;
    panelizationSummary: {
        containsEpubBubbles: boolean;
        containsImageBubbles: boolean;
    };
    imageLinks: {
        smallThumbnail: string;
        thumbnail: string;
    };
    language: string;
    previewLink: string;
    infoLink: string;
    canonicalVolumeLink: string;
};

export type BookOptions = {
    date: string;
    body?: string;
    bookIsbn: string;
    providers: string[];
};

export default async function getBook(
    options: BookOptions,
    fileName: string
): Promise<CleanBook[]> {
    const {bookIsbn, providers} = options;
    try {
        const book = (await isbn.provider(providers).resolve(bookIsbn)) as Book;
        exportVariable("BookTitle", book.title);
        return (await addBook(options, book, fileName)) as CleanBook[];
    } catch (error) {
        throw new Error(error.message);
    }
}
