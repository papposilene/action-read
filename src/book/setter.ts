import {exportVariable} from "@actions/core";
import cleanBook, {CleanBook} from "./cleaner";
import {sortByDate} from "../utils/globals";
import toJson from "../utils/json";
import {Book, BookOptions} from "./getter";

export default async function addBook(
    options: BookOptions,
    book: Book,
    fileName: string
) {
    // convert the file to JSON
    const readListJson = (await toJson(fileName)) as CleanBook[];
    // clean up book data
    const newBook: CleanBook = cleanBook(options, book);
    // export book thumbnail to download later
    if (newBook.imageLinks && newBook.imageLinks.thumbnail) {
        exportVariable("BookThumbOutput", `book-${newBook.isbn}.png`);
        exportVariable("BookThumb", newBook.imageLinks.thumbnail);
    }
    // append new book
    readListJson.push(newBook);
    return sortByDate(readListJson);
}
