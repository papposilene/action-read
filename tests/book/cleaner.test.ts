import cleanBook from "../../src/book/cleaner";
import book from "../fixture.json";

const date = "2020-09-12";

it("cleanBook", () =>
    expect(
        cleanBook(
            {
                date,
                body: "I loved it!",
                bookIsbn: "0525658181",
                providers: []
            },

            book
        )
    ).toMatchSnapshot());
