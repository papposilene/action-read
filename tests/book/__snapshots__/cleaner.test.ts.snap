// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cleanBook 1`] = `
Object {
  "authors": Array [
    "Yaa Gyasi",
  ],
  "canonicalVolumeLink": "https://books.google.com/books/about/Transcendent_Kingdom.html?hl=&id=ty19yQEACAAJ",
  "categories": Array [
    "Fiction",
  ],
  "dateFinished": "2020-09-12",
  "description": "A novel about faith, science, religion, and family that tells the deeply moving portrait of a family of Ghanaian immigrants ravaged by depression and addiction and grief, narrated by a fifth year candidate in neuroscience at Stanford school of medicine studying the neural circuits of reward seeking behavior in miceâ€¦",
  "imageLinks": Object {
    "smallThumbnail": "https://books.google.com/books/content?id=ty19yQEACAAJ&printsec=frontcover&img=1&zoom=5&source=gbs_api",
    "thumbnail": "https://books.google.com/books/content?id=ty19yQEACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api",
  },
  "industryIdentifiers": Array [
    Object {
      "identifier": "0525658181",
      "type": "ISBN_10",
    },
    Object {
      "identifier": "9780525658184",
      "type": "ISBN_13",
    },
  ],
  "isbn": "0525658181",
  "language": "en",
  "notes": "I loved it!",
  "pageCount": 288,
  "printType": "BOOK",
  "publishedDate": "2020",
  "title": "Transcendent Kingdom",
}
`;
